src/lib/tuning/tuner.hpp

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "scheduler/abstract_task.hpp"
+#include "tuning/abstract_evaluator.hpp"
+#include "tuning/abstract_selector.hpp"
+
+namespace opossum {
+
+/**
+ * A Tuner encapsulates the process of analyzing the current system state and
+ * performing specific modifications to optimize the systems performance.
+ *
+ * It uses AbstractEvaluators to generate TuningChoices which represent possible
+ * modifications together with their expected performance impact and costs.
+ * TuningChoices are transformed into a concrete operation sequence
+ * by an AbstractSelector, that also considers a cost budget that may not be
+ * exceeded at any point in the sequence.
+ *
+ * While executing the operation sequence, the Tuner considers a time budget
+ * and stops the execution once it exceeds that budget.
+ *
+ * Both the time and cost budget (the latter being enforced by the AbstractSelector)
+ * are initially infinite (i.e. disabled).
+ */
+class Tuner {
+ public:
+  using Runtime = std::chrono::duration<float, std::chrono::seconds::period>;
+  using RuntimeClock = std::chrono::high_resolution_clock;
+
+  enum class Status { Unknown, Running, Completed, EvaluationTimeout, SelectionTimeout, ExecutionTimeout, Timeout };
+
+  static constexpr float NoBudget = std::numeric_limits<float>::infinity();
+
+  Tuner();
+
+  /**
+   * Accessors for the AbstractEvaluators used for the tuning process.
+   *
+   * The Tuner takes ownership of the supplied evaluator. The same evaluator
+   * instance may never be used in more than one Tuner.
+   */
+  void add_evaluator(std::unique_ptr<AbstractEvaluator>&& evaluator);
+  void remove_evaluator(std::size_t index);
```

Comment:
When would I remove a tuner?

Reply:
As a particular Tuner can be scheduled multiple times, it might make sense to re-configure it between runs. Although this is unlikely, it can make sense to remove one evaluator - e.g. if the Tuner keeps running out of time repeatedly.
If you think that this is not needed, we could remove it.

Reply:
Okay, we keep it for now